apps/src/applab/RecordListener.js

```diff
@@ -0,0 +1,230 @@
+'use strict';
+
+/**
+ * Class which allows callers to listen to changes to data tables. This is currently
+ * done by polling. Some modest rate-limiting is achieved by waiting for one second
+ * after the previous response is processed for data from a given table, rather than
+ * polling on a fixed interval and allowing a backlog to accumulate.
+ *
+ * The most logical incremental performance improvement is to use pusher to only
+ * re-read from the table if any of its contents have changed. Down the road, we
+ * may be able to use DynamoDB Streams / Lambda / SQS to deliver fine-grained
+ * notifications to the client without ever needing to  re-read the entire table.
+ *
+ * @constructor
+ */
+var RecordListener = module.exports = function () {
+  /**
+   * Map from table name to table handler.
+   * @private {Object.<string, TableHandler>}
+   */
+  this.tableHandlers_ = {};
+};
+
+/**
+ * Number of ms to wait before polling each table for more data after finishing
+ * processing the last batch of data for that table.
+ * @type {number}
+ */
+var RECORD_INTERVAL = 1000; // 1 second
+
+var EventType = {
+  CREATE: 'create',
+  UPDATE: 'update',
+  DELETE: 'delete'
+};
+RecordListener.EventType = EventType;
+
+/**
+ * Adds a listener which calls the callback when changes are made to the table.
+ * The callback is called once per changed record.
```

Comment:
"addListener" methods typically supports multiple listeners.

Should we allow developers to add more than one listener on the same event?  This reduces coupling for larger apps because it allows components to listen for events without having to know about other components that are interested in the same event.  

If we do want to enforce the single listener, we should document that explicitly (not just in the returns clause) and perhaps call this "setListener" or "registerListener" to suggest that only a single listener is supported.